internal/pkg/fieldtags/analyzer.go

```diff
@@ -0,0 +1,110 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package fieldtags defines an analyzer that identifies struct fields tagged with datapol annotations
+// indicating that they contain SPII.
+package fieldtags
+
+import (
+	"go/ast"
+	"reflect"
+	"strings"
+
+	"golang.org/x/tools/go/analysis"
+	"golang.org/x/tools/go/analysis/passes/inspect"
+	"golang.org/x/tools/go/analysis/passes/structtag"
+	"golang.org/x/tools/go/ast/inspector"
+)
+
+type keyValue struct {
+	key   string
+	value string
+}
+
+type sourcePatterns []keyValue
+
+var patterns sourcePatterns = []keyValue{
+	{"levee", "source"},
+}
+
+var Analyzer = &analysis.Analyzer{
+	Name:       "fieldtags",
+	Doc:        "This analyzer identifies Source fields based on their tags.",
+	Run:        run,
+	Requires:   []*analysis.Analyzer{inspect.Analyzer, structtag.Analyzer},
+	ResultType: reflect.TypeOf(new([]*ast.Field)).Elem(),
+}
+
+func run(pass *analysis.Pass) (interface{}, error) {
+	inspect := pass.ResultOf[inspect.Analyzer].(*inspector.Inspector)
+	results := []*ast.Field{}
+
+	nodeFilter := []ast.Node{
+		(*ast.Field)(nil),
+	}
+
+	inspect.Preorder(nodeFilter, func(n ast.Node) {
+		field, ok := n.(*ast.Field)
+		if !ok {
+			return
+		}
+		if patterns.isSource(field) {
+			results = append(results, field)
+			pass.Reportf(field.Pos(), "tagged field")
+		}
+	})
+	return results, nil
+}
+
+func (ps *sourcePatterns) isSource(field *ast.Field) bool {
+	if field.Tag == nil {
+		return false
+	}
+
+	tag := field.Tag.Value
+
+	i := 1 // skip backtick
+	j := 1
+	for j < len(tag) {
```

Comment:
Could this be refactored to a regex?

Reply:
Here's a stab at implementing your suggestion : https://github.com/mlevesquedion/go-flow-levee/compare/field-tags-analyzer...mlevesquedion:refactor-field-tags-matching-into-regex

I think the regex version may be more readable and potentially more robust as well.

It seems I can't reference the current PR from the PR I linked above.

Reply:
I am totally fine with just a TODO here, and follow up with another PR to refactor to regex.

Reply:
SG. Done.